### PR TITLE
#171648173 Feature allows users to sign in using Google and Facebook

### PR DESCRIPTION
**Pull request by Irutingabo**
#### Description
A feature allows users to sign-in using their Google account(Gmail) and Facebook account
#### Type of change

Please select the relevant option

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Others (cosmetics, styling, improvements)
- [ ] This change requires a documentation update

#### How Has This Been Tested?

- [X] Unit
- [ ] Integration
- [ ] End-to-end
- [ ] No testing required

#### How to Test
Sign in using existing Gmail or Facebook account
Give instructions on how reviewers can verify your work

#### Any background context you want to add (Optional)

Screenshots or GIFs

#### STORY ID

[Finishes #171648173]

### CODE REVIEW

**Review by brianineza01 — APPROVED** [review]

**Review by Agnes-N — APPROVED** [review]

File: server/controllers/authController.js [hunk 1]
```diff
@@ -114,13 +114,52 @@ class AuthController {
...
               isVerified
             }
           }
         }));
       },
     )(req, res);
   }
+
+  /**
+     * This method handle the sign request.
```
Comment (placideirandora) [inline server/controllers/authController.js]: Update the comment. This is a Social Sign In request instead

File: server/controllers/authController.js [hunk 2]
```diff
@@ -114,13 +114,52 @@ class AuthController {
...
+     * @param {object} req The user's request.
+     * @param {object} res The response.
+     * @returns {object} The status and some data of the user.
+     */
+  static async socialAuth(req, res) {
+    const {
+      id, firstname, lastname, email, role, isVerified, lineManager
+    } = req.user;
```
Comment (placideirandora) [inline server/controllers/authController.js]: I don’t see a generated username as well. By signing in with social platforms, If you don’t have access to the username, you should use their first name as their username.

File: server/controllers/authController.js [hunk 3]
```diff
@@ -114,13 +114,52 @@ class AuthController {
...
+
+    const token = await AuthHelper.createToken({
+      id,
+      firstname,
+      lastname,
+      lineManager,
+      email,
+      role,
+      isVerified
+    });
```
Comment (placideirandora) [inline server/controllers/authController.js]: The username should be included as well.

File: server/controllers/authController.js [hunk 4]
```diff
@@ -114,13 +114,52 @@ class AuthController {
...
+      data: {
+        token,
+        user: {
+          firstname,
+          lastname,
+          email,
+          role,
+          lineManager,
+          isVerified
+        }
```
Comment (placideirandora) [inline server/controllers/authController.js:192]: Specify the username as well.

File: server/helpers/verifyUser.js [hunk 5]
```diff
@@ -25,5 +27,53 @@ class VerifyUser {
     }
     return done('password or email is incorrect');
   }
+
+
+  /**
+     * Retrieve hashed the password.
```
Comment (placideirandora) [inline server/helpers/verifyUser.js]: Update this comment. **the** should come before hashed.

File: server/helpers/verifyUser.js [hunk 6]
```diff
@@ -25,5 +27,53 @@ class VerifyUser {
...
+     * @param {string} aT The user's token offered by google to be checked.
+     * @param {string} rT The user's token offered by google to be checked.
+     * @param {string} profile The user info
+     * @param {string} done function supplying user credentials after user is successfully checked.
+     * @returns {boolean} message or User details after passing all checks.
+     */
+  static async inSignSocial(aT, rT, profile, done) {
```
Comment (placideirandora) [inline server/helpers/verifyUser.js]: Update the method name to **socialSignIn**

File: server/middlewares/checkUser.js [hunk 7]
```diff
@@ -46,5 +46,35 @@ class checkUser {
...
+  static async isAuthWithGoogle(req, res, next) {
+    passport.authenticate('google', { failureRedirect: '/google', session: false }, (err, user) => {
+      if (err) { return next(err); }
+      req.user = user;
+      return next();
+    })(req, res, next);
+  }
+
+  /**
+     * This method Checks if User is sucessfully authenticated with google.
```
Comment (placideirandora) [inline server/middlewares/checkUser.js]: Update the comment. This is for Facebook instead.

File: server/tests/auth-signin.test.js [hunk 8]
```diff
@@ -4,7 +4,7 @@ import { describe, it } from 'mocha';
 import chaiHttp from 'chai-http';
```
Comment (placideirandora) [inline server/tests/auth-signin.test.js]: Remove this empty space.

File: server/tests/mochData/socialMock.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+const googleProfile = {
+  provider: 'google',
+  id,
+  displayName: `${firstName} ${lastName}`,
+  name: { familyName: lastName, givenName: firstName },
+  emails: [{ value: email, type: 'ACCOUNT' }],
+  photos: [
+    {
+      value: 'https://lh3.googleusercontent.com/a-/AAuE7mD7hIQTg4_0WvV34It93y5tEeDT87JbKRW5Aln_=s50'
```
Comment (placideirandora) [inline server/tests/mochData/socialMock.js]: Put a fake value here as well. No need to provide real one. You can replace the content from **.com/** to any invalid url/data.

File: server/tests/mochData/socialMock.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  ],
+  _json: {
+    id,
+    email,
+    name: { familyName: lastName, givenName: firstName },
+    displayName: `${firstName} ${lastName}`,
+    image: {
+      url: 'https://lh3.googleusercontent.com/a-/AAuE7mD7hIQTg4_0WvV34It93y5tEeDT87JbKRW5Aln_=s50'
```
Comment (placideirandora) [inline server/tests/mochData/socialMock.js]: Here as well.

File: server/tests/mochData/socialMock.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+const facebookProfile = {
+  provider: 'facebook',
+  id,
+  displayName: `${firstName} ${lastName}`,
+  name: { familyName: lastName, givenName: firstName },
+  emails: [{ value: email, type: 'ACCOUNT' }],
+  photos: [
+    {
+      value: 'https://lh3.googleusercontent.com/a-/AAuE7mD7hIQTg4_0WvV34It93y5tEeDT87JbKRW5Aln_=s50'
```
Comment (placideirandora) [inline server/tests/mochData/socialMock.js]: Here as well.

File: server/tests/mochData/socialMock.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  ],
+  _json: {
+    id,
+    email,
+    name: { familyName: lastName, givenName: firstName },
+    displayName: `${firstName} ${lastName}`,
+    image: {
+      url: 'https://lh3.googleusercontent.com/a-/AAuE7mD7hIQTg4_0WvV34It93y5tEeDT87JbKRW5Aln_=s50'
```
Comment (placideirandora) [inline server/tests/mochData/socialMock.js]: Here as well.

File: server/tests/mochData/socialMockData.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
+const mockData = [
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: So, what’s the difference between **socialMock** and **socialMockData**? You should use a single efficient approach between them.

File: server/tests/mochData/socialMockData.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  {
+    username: 'Patsy Waters',
+    firstname: 'Patsy',
+    lastname: 'Waters',
+    email: 'Anais79@hotmail.com',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: I hope that this email does not exist. If so, usea fake email. No need to provide real data in testing and that’s why It’s called, **Mocked Data**. We just follow the desired structure of information for testing purpose.

File: server/tests/mochData/socialMockData.js [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    role: 'Requester',
+    isVerified: true,
+    googleId: 'f02369b8-0081-4691-a732-516e59994234',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: Change this ID into a random fake one.

File: server/tests/mochData/socialMockData.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    isGoogleSigned: true
+  },
+  {
+    id: '105602922558884032675',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: This one as well.

File: server/tests/mochData/socialMockData.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    username: 'Champs Dev',
+    firstname: 'Champs',
+    lastname: 'Dev',
+    email: 'champsdev2@gmail.com',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: This email is valid. Use an invalid email instead.

File: server/tests/mochData/socialMockData.js [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    role: 'Requester',
+    isVerified: true,
+    googleId: 'f02369b8-0081-4691-a732-516e59994234',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: Here as well.

File: server/tests/mochData/socialMockData.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    isGoogleSigned: true
+  },
+  {
+    id: 1,
+    googleId: '105602922558884032675',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: Here as well.

File: server/tests/mochData/socialMockData.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    firstname: 'Champs',
+    lastname: 'Dev',
+    email: 'champsdev2@gmail.com',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: Here as well.

File: server/tests/mochData/socialMockData.js [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    role: 'Requester',
+    isVerified: true,
+    isGoogleSigned: true,
+    facebookId: null,
+    password: null,
+    isFacebookSigned: null
+  },
+  {
+    email: 'champsdev2@gmail.com',
```
Comment (placideirandora) [inline server/tests/mochData/socialMockData.js]: Here as well.

File: server/tests/oauth.test.js [hunk 22]
```diff
@@ -0,0 +1,112 @@
+import passport from 'passport';
+import * as chai from 'chai';
+import sinonChai from 'sinon-chai';
+import sinon from 'sinon';
+import request from 'supertest';
+import { googleProfile, facebookProfile } from './mochData/socialMock';
+import data from './mochData/socialMockData';
+import VerifyUser from '../helpers/verifyUser';
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Use **Camel Case** instead

File: server/tests/oauth.test.js [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+import CheckUser from '../middlewares/checkUser';
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Use **Camel Case** here as well.

File: server/tests/oauth.test.js [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+chai.should();
+const { expect } = chai;
+
+
+describe('Social authentication tests', () => {
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  it('should return user data when it get the profile from the social account', async () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **should return a user profile from Google**.

File: server/tests/oauth.test.js [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+import CheckUser from '../middlewares/checkUser';
+
+import app from '../index';
+
+chai.use(sinonChai);
+chai.should();
+const { expect } = chai;
+
+
+describe('Social authentication tests', () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js:20]: Update the description to **Google Social Authentication Test**.

File: server/tests/oauth.test.js [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+  it('should return user data when it get the profile from the social account', async () => {
+    const accessToken = 'zzzz-zzzz-zzzz';
+    const refreshToken = 'zzzzzzz';
+    const callBack = sinon.spy();
+    await VerifyUser.inSignSocial(accessToken, refreshToken, googleProfile, callBack);
+    expect(callBack.withArgs(null, googleProfile));
+  });
+});
+
+describe('Social authentication tests', () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the comment to **Facebook Social Authentication Test**

File: server/tests/oauth.test.js [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  it('should return user data when it get the profile from the social account', async () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **should return a user profile from Facebook**.

File: server/tests/oauth.test.js [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+    const accessToken = 'zzzz-zzzz-zzzz';
+    const refreshToken = 'zzzzzzz';
+    const callBack = sinon.spy();
+    await VerifyUser.inSignSocial(accessToken, refreshToken, facebookProfile, callBack);
+    expect(callBack.withArgs(null, facebookProfile));
+  });
+});
+
+describe('User sign in successfully with google', () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **Google Social Sign In Test**

File: server/tests/oauth.test.js [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+  before(async () => {
+    sinon.stub(passport, 'authenticate').callsFake((strategy, options, callback) => {
+      callback(null, data[3], null);
+      return () => { };
+    });
+  });
+  after(() => {
+    passport.authenticate.restore();
+  });
+  it('should sign in user', (done) => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **should successfully sign in the user with Google**

File: server/tests/oauth.test.js [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+        expect(res.body).to.be.a('object');
+        expect(res.body).to.have.property('status');
+        expect(res.body).to.have.property('message');
+        done();
+      });
+  });
+});
+
+
+describe('User sign in successfully with facebook', () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **Facebook Social Sign In Test**

File: server/tests/oauth.test.js [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+  before(async () => {
+    sinon.stub(passport, 'authenticate').callsFake((strategy, options, callback) => {
+      callback(null, data[3], null);
+      return () => { };
+    });
+  });
+  after(() => {
+    passport.authenticate.restore();
+  });
+  it('should sign in user', (done) => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **should successfully sign in the user with Facebook**

File: server/tests/oauth.test.js [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+        expect(res).to.have.status(200);
+        expect(res.body).to.be.a('object');
+        expect(res.body).to.have.property('status');
+        expect(res.body).to.have.property('message');
+        done();
+      });
+  });
+});
+
+describe('Authentication fails to return user google profile', () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **should redirect the user to the Google Auth endpoint once not already authenticated**

File: server/tests/oauth.test.js [hunk 33]
```diff
@@ -0,0 +1,112 @@
...
+  it('should call next(error) once', () => {
+    const nextSpy = sinon.spy();
+
+    CheckUser.isAuthWithGoogle({}, {}, nextSpy);
+    expect(nextSpy.calledOnce).to.equal(true);
+  });
+});
+
+
+describe('Authentication fails to return user facebook profile', () => {
```
Comment (placideirandora) [inline server/tests/oauth.test.js]: Update the description to **should redirect the user to the Facebook Auth endpoint once not already authenticated**

**Review by placideirandora — CHANGES_REQUESTED** [review]
Implement the Feedback.

File: server/controllers/authController.js (hunk 1 above)
Comment (Irutingabo) [reply server/controllers/authController.js]: Thanks! Feedback Implemented.

File: server/controllers/authController.js (hunk 2 above)
Comment (Irutingabo) [reply server/controllers/authController.js]: Feedback Implemented!

File: server/controllers/authController.js (hunk 3 above)
Comment (Irutingabo) [reply server/controllers/authController.js]: Feedback Implemented!

File: server/controllers/authController.js (hunk 4 above)
Comment (Irutingabo) [reply server/controllers/authController.js:192]: Feedback Implemented!

File: server/helpers/verifyUser.js (hunk 5 above)
Comment (Irutingabo) [reply server/helpers/verifyUser.js]: Feedback Implemented!

File: server/helpers/verifyUser.js (hunk 6 above)
Comment (Irutingabo) [reply server/helpers/verifyUser.js]: Feedback Implemented!

File: server/middlewares/checkUser.js (hunk 7 above)
Comment (Irutingabo) [reply server/middlewares/checkUser.js]: Feedback Implemented!

File: server/tests/auth-signin.test.js (hunk 8 above)
Comment (Irutingabo) [reply server/tests/auth-signin.test.js]: Feedback Implemented!

File: server/tests/mochData/socialMock.js (hunk 9 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMock.js]: Feedback Implemented!

File: server/tests/mochData/socialMock.js (hunk 10 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMock.js]: Feedback Implemented!

File: server/tests/mochData/socialMock.js (hunk 11 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMock.js]: Feedback Implemented!

File: server/tests/mochData/socialMock.js (hunk 12 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMock.js]: Feedback Implemented!

File: server/tests/mochData/socialMockData.js (hunk 13 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented! FIle deleted

File: server/tests/mochData/socialMockData.js (hunk 14 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented! They are using fake emails now.

File: server/tests/mochData/socialMockData.js (hunk 16 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented!

File: server/tests/mochData/socialMockData.js (hunk 17 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented! using fake email now

File: server/tests/mochData/socialMockData.js (hunk 18 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented!

File: server/tests/mochData/socialMockData.js (hunk 19 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented!

File: server/tests/mochData/socialMockData.js (hunk 20 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented!

File: server/tests/mochData/socialMockData.js (hunk 21 above)
Comment (Irutingabo) [reply server/tests/mochData/socialMockData.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 22 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 23 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 24 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 25 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js:20]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 26 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 27 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 28 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 29 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 30 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 31 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 32 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 33 above)
Comment (Irutingabo) [reply server/tests/oauth.test.js]: Feedback Implemented!

File: server/tests/oauth.test.js (hunk 25 above)
Comment (placideirandora) [reply server/tests/oauth.test.js:20]: Not yet.

**Review by YankyJayChris — APPROVED** [review]

**Review by placideirandora — APPROVED** [review]

**Pull request merged**